src/lib/es5.d.ts

```diff
@@ -180,6 +180,18 @@ interface ObjectConstructor {
       * Prevents the modification of existing property attributes and values, and prevents the addition of new properties.
       * @param o Object on which to lock the attributes.
       */
+    freeze<T>(a: T[]): ReadonlyArray<T>;
+
+    /**
+      * Prevents the modification of existing property attributes and values, and prevents the addition of new properties.
+      * @param o Object on which to lock the attributes.
+      */
+    freeze<T>(f: (...args: any[]) => T): (...args: any[]) => T;

```

Comment:
@falsandtru I think the definition I suggested works for both classes and generic functions (but it doesn't solve the problem in #9366).